// react-form-with-constraints/lib-es5/Field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Field = void 0;
var clearArray_1 = require("./clearArray");
var FieldFeedbackType_1 = require("./FieldFeedbackType");
var Field = (function () {
    function Field(name) {
        this.name = name;
        this.validations = [];
    }
    Field.prototype.addOrReplaceValidation = function (validation) {
        var i = this.validations.findIndex(function (_validation) { return _validation.key === validation.key; });
        if (i > -1)
            this.validations[i] = validation;
        else
            this.validations.push(validation);
    };
    Field.prototype.clearValidations = function () {
        clearArray_1.clearArray(this.validations);
    };
    Field.prototype.hasFeedbacksOfType = function (type, fieldFeedbacksKey) {
        return this.validations.some(function (fieldFeedback) {
            return (fieldFeedbacksKey === undefined ||
                fieldFeedback.key.startsWith(fieldFeedbacksKey + ".")) &&
                fieldFeedback.type === type &&
                fieldFeedback.show === true;
        });
    };
    Field.prototype.hasErrors = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.FieldFeedbackType.Error, fieldFeedbacksKey);
    };
    Field.prototype.hasWarnings = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.FieldFeedbackType.Warning, fieldFeedbacksKey);
    };
    Field.prototype.hasInfos = function (fieldFeedbacksKey) {
        return this.hasFeedbacksOfType(FieldFeedbackType_1.FieldFeedbackType.Info, fieldFeedbacksKey);
    };
    Field.prototype.hasFeedbacks = function (fieldFeedbacksKey) {
        return (this.hasErrors(fieldFeedbacksKey) ||
            this.hasWarnings(fieldFeedbacksKey) ||
            this.hasInfos(fieldFeedbacksKey));
    };
    Field.prototype.isValid = function () {
        return !this.hasErrors();
    };
    return Field;
}());
exports.Field = Field;
