// react-form-with-constraints/lib-es5/withValidateFieldEventEmitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.withValidateFieldEventEmitter = exports.ValidateFieldEvent = void 0;
var EventEmitter_1 = require("./EventEmitter");
exports.ValidateFieldEvent = 'VALIDATE_FIELD_EVENT';
function withValidateFieldEventEmitter(Base) {
    return (function (_super) {
        __extends(ValidateFieldEventEmitter, _super);
        function ValidateFieldEventEmitter() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.validateFieldEventEmitter = new EventEmitter_1.EventEmitter();
            return _this;
        }
        ValidateFieldEventEmitter.prototype.emitValidateFieldEvent = function (input) {
            return this.validateFieldEventEmitter.emitAsync(exports.ValidateFieldEvent, input);
        };
        ValidateFieldEventEmitter.prototype.addValidateFieldEventListener = function (listener) {
            this.validateFieldEventEmitter.addListener(exports.ValidateFieldEvent, listener);
        };
        ValidateFieldEventEmitter.prototype.removeValidateFieldEventListener = function (listener) {
            this.validateFieldEventEmitter.removeListener(exports.ValidateFieldEvent, listener);
        };
        return ValidateFieldEventEmitter;
    }(Base));
}
exports.withValidateFieldEventEmitter = withValidateFieldEventEmitter;
