// react-form-with-constraints/lib-es5/withFieldWillValidateEventEmitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.withFieldWillValidateEventEmitter = exports.FieldWillValidateEvent = void 0;
var EventEmitter_1 = require("./EventEmitter");
exports.FieldWillValidateEvent = 'FIELD_WILL_VALIDATE_EVENT';
function withFieldWillValidateEventEmitter(Base) {
    return (function (_super) {
        __extends(FieldWillValidateEventEmitter, _super);
        function FieldWillValidateEventEmitter() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.fieldWillValidateEventEmitter = new EventEmitter_1.EventEmitter();
            return _this;
        }
        FieldWillValidateEventEmitter.prototype.emitFieldWillValidateEvent = function (fieldName) {
            return this.fieldWillValidateEventEmitter.emitSync(exports.FieldWillValidateEvent, fieldName);
        };
        FieldWillValidateEventEmitter.prototype.addFieldWillValidateEventListener = function (listener) {
            this.fieldWillValidateEventEmitter.addListener(exports.FieldWillValidateEvent, listener);
        };
        FieldWillValidateEventEmitter.prototype.removeFieldWillValidateEventListener = function (listener) {
            this.fieldWillValidateEventEmitter.removeListener(exports.FieldWillValidateEvent, listener);
        };
        return FieldWillValidateEventEmitter;
    }(Base));
}
exports.withFieldWillValidateEventEmitter = withFieldWillValidateEventEmitter;
