// react-form-with-constraints/lib-es5/withFieldDidResetEventEmitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.withFieldDidResetEventEmitter = exports.FieldDidResetEvent = void 0;
var EventEmitter_1 = require("./EventEmitter");
exports.FieldDidResetEvent = 'FIELD_DID_RESET_EVENT';
function withFieldDidResetEventEmitter(Base) {
    return (function (_super) {
        __extends(ResetFieldEvenEmitter, _super);
        function ResetFieldEvenEmitter() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.fieldDidResetEventEmitter = new EventEmitter_1.EventEmitter();
            return _this;
        }
        ResetFieldEvenEmitter.prototype.emitFieldDidResetEvent = function (field) {
            return this.fieldDidResetEventEmitter.emitSync(exports.FieldDidResetEvent, field);
        };
        ResetFieldEvenEmitter.prototype.addFieldDidResetEventListener = function (listener) {
            this.fieldDidResetEventEmitter.addListener(exports.FieldDidResetEvent, listener);
        };
        ResetFieldEvenEmitter.prototype.removeFieldDidResetEventListener = function (listener) {
            this.fieldDidResetEventEmitter.removeListener(exports.FieldDidResetEvent, listener);
        };
        return ResetFieldEvenEmitter;
    }(Base));
}
exports.withFieldDidResetEventEmitter = withFieldDidResetEventEmitter;
