// react-form-with-constraints/lib-es5/withFieldDidValidateEventEmitter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.withFieldDidValidateEventEmitter = exports.FieldDidValidateEvent = void 0;
var EventEmitter_1 = require("./EventEmitter");
exports.FieldDidValidateEvent = 'FIELD_DID_VALIDATE_EVENT';
function withFieldDidValidateEventEmitter(Base) {
    return (function (_super) {
        __extends(FieldDidValidateEventEmitter, _super);
        function FieldDidValidateEventEmitter() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.fieldDidValidateEventEmitter = new EventEmitter_1.EventEmitter();
            return _this;
        }
        FieldDidValidateEventEmitter.prototype.emitFieldDidValidateEvent = function (field) {
            return this.fieldDidValidateEventEmitter.emitSync(exports.FieldDidValidateEvent, field);
        };
        FieldDidValidateEventEmitter.prototype.addFieldDidValidateEventListener = function (listener) {
            this.fieldDidValidateEventEmitter.addListener(exports.FieldDidValidateEvent, listener);
        };
        FieldDidValidateEventEmitter.prototype.removeFieldDidValidateEventListener = function (listener) {
            this.fieldDidValidateEventEmitter.removeListener(exports.FieldDidValidateEvent, listener);
        };
        return FieldDidValidateEventEmitter;
    }(Base));
}
exports.withFieldDidValidateEventEmitter = withFieldDidValidateEventEmitter;
